File: 0207-course-schedule/0207-course-schedule.js
/**
 * @param {number} numCourses
 * @param {number[][]} prerequisites
 * @return {boolean}
 */
var canFinish = function(numCourses, prerequisites) {
    const visited = new Set()
    const preReq = new Array(numCourses).fill().map(() => [])

    for(const [crs, pre] of prerequisites){
        preReq[crs].push(pre)
    }

    function dfs(crs){
        if(visited.has(crs)){
            return false
        }
        if(preReq[crs] == []){
            return true
        }

        visited.add(crs)
        for(const pre of preReq[crs]){
            if(!dfs(pre)){
                return false
            }
        }
        visited.delete(crs)
        preReq[crs] = []
        return true
    }    

    for(let i=0; i < numCourses; i++){
        if(!dfs(i)){
            return false
        }
    }
    return true
};

File: 0207-course-schedule/0207-course-schedule.py
class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        adjList = defaultdict(list)
        for c, prereq in prerequisites:
            adjList[c].append(prereq)

        taken = set()
        taking = set()

        def dfs(course):
            if course in taking:
                return False
            if course in taken:
                return True
            
            taking.add(course)
            
            for prereq in adjList[course]:
                if not dfs(prereq):
                    return False

            taking.remove(course)
            taken.add(course)
            return True


        for i in range(numCourses):
            if not dfs(i):
                return False
        return True